### PR TITLE
[FILESYSTEMS] Fix pool memory disclosure in filesystem drivers supporting FS_INFORMATION_CLASS.FileFsVolumeInformation

### PR DESCRIPTION
**Pull request by khanhnt2**
## Purpose

Fix structure alignment leads to pool memory disclosure in drivers support `FS_INFORMATION_CLASS.FileFsVolumeInformation`

## Analysis

The structure `_FILE_FS_VOLUME_INFORMATION`

```
kd> dt _FIlE_FS_VOLUME_INFORMATION
shell32!_FILE_FS_VOLUME_INFORMATION
   +0x000 VolumeCreationTime : _LARGE_INTEGER
   +0x008 VolumeSerialNumber : Uint4B
   +0x00c VolumeLabelLength : Uint4B
   +0x010 SupportsObjects  : UChar
   +0x012 VolumeLabel      : [1] Wchar
```

is added 1 padding byte after the field `SupportsObjects` so it will lead to memory disclosure although all fields are initialized.

The backtrace report of bochspwn-reloaded:

```
 #0  0x80581cb8 ((00181cb8) ntoskrnl.exe!memcpy+48)
 #1  0x80485686 ((00085686) ntoskrnl.exe!IopCompleteRequest+1c6 [h:\project\reactos\ntoskrnl\io\iomgr\irp.c @ 292])
 #2  0x8047b3aa ((0007b3aa) ntoskrnl.exe!IopPerformSynchronousRequest+ba [h:\project\reactos\ntoskrnl\io\iomgr\iofunc.c @ 158])
 #3  0x8047dd80 ((0007dd80) ntoskrnl.exe!NtQueryVolumeInformationFile+6c0 [h:\project\reactos\ntoskrnl\io\iomgr\iofunc.c @ 4341])
 #4  0x8054dc5b ((0014dc5b) ntoskrnl.exe!KiSystemCallTrampoline+1b [h:\project\reactos\ntoskrnl\include\internal\i386\ke.h @ 766])
 #5  0x8054bac8 ((0014bac8) ntoskrnl.exe!KiSystemServiceHandler+278 [h:\project\reactos\ntoskrnl\ke\i386\traphdlr.c @ 1846])
 #6  0x80403ea5 ((00003ea5) ntoskrnl.exe!KiFastCallEntry+96)
```

This bug affects all filesystem drivers support `FS_INFORMATION_CLASS.FileFsVolumeInformation`. Luckily some of them are zeroed system memory before using so I only zeroing the rest.

For btrfs driver, I think I should pull request to it instead of fixing this bug in ReactOS.

### CODE REVIEW

File: drivers/filesystems/cdfs/volinfo.c [hunk 1]
```diff
@@ -160,6 +160,8 @@ Return Value:
         //  and false if it couldn't wait for any I/O to complete.
         //
 
+        RtlZeroMemory(Irp->AssociatedIrp.SystemBuffer, Length);
```
Comment (Extravert-ir) [inline drivers/filesystems/cdfs/volinfo.c:163]: Hm, does it mean Windows also has this issue? :)

Comment (khanhnt2) [reply drivers/filesystems/cdfs/volinfo.c:163]: Which issue? Structure alignment or using RtlZeroMemory to clear memory?

Comment (Extravert-ir) [reply drivers/filesystems/cdfs/volinfo.c:163]: I mean cdfs driver is taken from Microsoft samples repository

Comment (khanhnt2) [reply drivers/filesystems/cdfs/volinfo.c:163]: No, I haven't. I don't know cdfs driver is taken from Microsoft.

**Comment by Extravert-ir** [comment]
By the way: I'm not quite familiar with that security techniques, but isn't it better to allocate such a buffer on stack and then just copy the whole one?

**Comment by binarymaster** [comment]
Related Btrfs issue addressed upstream: https://github.com/maharmstone/btrfs/commit/30f4f0fef0a99e44471e878fad1dd7b2b9fb6c0a

File: drivers/filesystems/cdfs/volinfo.c (hunk 1 above)
Comment (Saibamen) [reply drivers/filesystems/cdfs/volinfo.c:163]: So please create PR for MS repo https://github.com/microsoft/Windows-driver-samples

**Comment by khanhnt2** [comment]
> 
> 
> By the way: I'm not quite familiar with that security techniques, but isn't it better to allocate such a buffer on stack and then just copy the whole one?

If you don't zero memory on stack, the bug will be still there.

File: drivers/filesystems/nfs/nfs41_driver.c [hunk 2]
```diff
@@ -4546,6 +4546,8 @@ NTSTATUS nfs41_QueryVolumeInformation(
     status = check_nfs41_dirquery_args(RxContext);
     if (status) goto out;
 
+    RtlZeroMemory(RxContext->Info.Buffer, RxContext->Info.LengthRemaining);
```
Comment (Extravert-ir) [inline drivers/filesystems/nfs/nfs41_driver.c:4549]: Please, wrap this change with `#ifdef __REACTOS__` like it's done in other places here

File: drivers/filesystems/npfs/volinfo.c [hunk 3]
```diff
@@ -78,7 +74,6 @@ NpQueryFsDeviceInfo(IN PVOID Buffer,
     PFILE_FS_DEVICE_INFORMATION InfoBuffer = Buffer;
     TRACE("Entered\n");
 
-    InfoBuffer->DeviceType = 0;
     InfoBuffer->Characteristics = 0;
```
Comment (Extravert-ir) [inline drivers/filesystems/npfs/volinfo.c:77]: If `InfoBuffer->DeviceType` assignment is removed, this one should be removed as well.

File: drivers/filesystems/cdfs/volinfo.c [hunk 4]
```diff
@@ -356,7 +358,6 @@ Return Value:
 
     Buffer->TotalAllocationUnits.QuadPart = LlSectorsFromBytes( Vcb->VolumeDasdFcb->AllocationSize.QuadPart );
 
-    Buffer->AvailableAllocationUnits.QuadPart = 0;
```
Comment (Extravert-ir) [inline drivers/filesystems/cdfs/volinfo.c:359]: This is an imported driver so I'd rather reduce the diff here

File: drivers/filesystems/cdfs/volinfo.c (hunk 1 above)
Comment (Extravert-ir) [inline drivers/filesystems/cdfs/volinfo.c:163]: Please, wrap this change with `#ifdef __REACTOS__`

**Review by Extravert-ir — CHANGES_REQUESTED** [review]

File: drivers/filesystems/nfs/nfs41_driver.c (hunk 2 above)
Comment (khanhnt2) [reply drivers/filesystems/nfs/nfs41_driver.c:4550]: Thank you. I fixed.

File: drivers/filesystems/npfs/volinfo.c (hunk 3 above)
Comment (khanhnt2) [reply drivers/filesystems/npfs/volinfo.c]: Thank you. I fixed.

File: drivers/filesystems/cdfs/volinfo.c (hunk 4 above)
Comment (khanhnt2) [reply drivers/filesystems/cdfs/volinfo.c]: Thank you. I reverted it.

File: drivers/filesystems/cdfs/volinfo.c (hunk 1 above)
Comment (khanhnt2) [reply drivers/filesystems/cdfs/volinfo.c:164]: Thank you, my mistake :D

Comment (khanhnt2) [reply drivers/filesystems/cdfs/volinfo.c:164]: I created a pull request to Microsoft repository: https://github.com/microsoft/Windows-driver-samples/pull/536

File: drivers/filesystems/nfs/nfs41_driver.c (hunk 2 above)
Comment (Saibamen) [reply drivers/filesystems/nfs/nfs41_driver.c:4550]: @khanhnt2: Do you think this line of code should also be added as PR in https://github.com/microsoft/Windows-driver-samples?

Comment (khanhnt2) [reply drivers/filesystems/nfs/nfs41_driver.c:4550]: @Saibamen I couldn't find nfs driver in Microsoft samples. Could you give me a link, I'm pleasure to do that.

Comment (Saibamen) [reply drivers/filesystems/nfs/nfs41_driver.c:4550]: Sorry, this file is from University of Michigan, not Microsoft.

**Pull request merged**